### ISSUE TITLE
Proposal: Unicode character classes in pmatch

### ISSUE THREAD

**Issue by Traubert** [hfst/hfst#560]
I'm proposing to add some runtime Unicode functionality to `hfst-pmatch`.

I'd like to change nothing on the pmatch-the-language side of things, except to introduce a new variable, `use-character-classes`, which would default to `off`. When set to `on`, this would cause symbols like `Alpha` to emit new special symbols, which `hfst-pmatch` would interpret using `libicu`.

This would break no existing workflows, but would cause new `pmatch` archives to eventually be built which wouldn't work correctly on pre-this-change versions of `hfst`.

At first, I'd implement the relatively simple

`Alpha`
`UppercaseAlpha`
`LowercaseAlpha`
`Num`
`Whitespace`

I'm not sure if there's any obviously correct way to handle `Punct`.

Then, for the casing functions like `UpCase()`, which has been a desideratum for a long time ( :) ), we could add starting new opening and closing special symbols, or possibly just endtag-style tags (eg. `@PMATCH_UPCASE_END@`), all text between which would be recased on the output side. This is much more likely to lead to bugs, but would obviously be very nice to have.

Any thoughts? I made a [branch](https://github.com/hfst/hfst/tree/pmatch-unicode-character-classes) with support for just `Alpha` for now:

```
$ echo 'set need-separators off set use-character-classes on regex Alpha;' | hfst-pmatch2fst | hfst-fst2txt
0	1	@UNICODE_ALPHA@	@UNICODE_ALPHA@	0.000000
1	0.000000
$ echo 'set need-separators off set use-character-classes on regex Alpha EndTag(alpha);' | hfst-pmatch2fst > /tmp/alpha.hfst
$ hfst-pmatch /tmp/alpha.hfst 
a
<alpha>a</alpha>

1
1

qqq
<alpha>q</alpha><alpha>q</alpha><alpha>q</alpha>

ä
<alpha>ä</alpha>

µ
<alpha>µ</alpha>
```

Now that I think of it, it would be more consistent naming to have `@PMATCH_UNICODE_ALPHA@`. Or maybe there should be more of a HFST-wide naming pattern for things like this?

**Comment by flammie** [comment]
> One very important thing to understand about character classes like this is that they don't work with respect to eg. composition and subtraction (like many other pmatch features of course, like context conditions). Pmatch in general should be conceived of as a final stage, the output of which should undergo finite-state algebra only when you know exactly what you're doing. That could be extensively warned about, but what isn't warned about is what happens inside pmatch scripts, eg. when you try to subtract from something that isn't really subtractable. I've added a bunch of warnings to help with that, and a rudimentary "call graph" during compilation to help seeing where things are coming from. Currently they are only triggered under --verbose, but it would not be unreasonable to print them always.

Yes this is definitely a good point, while I can kind of guess what can go wrong with using these features with composition etc., it is still hard to debug, and I can guess many users won't even guess that. Perhaps it also affects the perception that pmatch scripting is based on xerox' interactive user interface syntax, which a lot of users took as a full on scripting language solution.

> 
> Some sort of survival guide / cookbook is probably in order to help with this, but it's already good to have those extra warnings in place.

A cookbook would certainly be a good idea, also just a reference manual really with list of functions and limitations. 

> Because merging the branch would be safe in principle, since by default nothing changes, I might do that, but I now wonder if there should be a way to use both the old, subtractable latin-1 character classes and the new, nonsubtractable unicode character classes side-by-side. Also `use-character-classes` is maybe not the best name. `use-runtime-unicode-library-character-classes` would be a mouthful.

We could potentially have both with different syntaxes, after all, other regex dialects have various ways to do this too, so it is something people would expect. 

> (I've also started writing a "arc compressor". Especially prior to the existence of unicode arcs, but to some extent still, a huge proportion of arcs in transducers could be replaced with lists. Especially in the case of eg. `[? - A]`, which some rule writers are fond of using, and when used in a nested way, can result in almost-sigma being pasted all over the place. Lists are not subtraction-safe in rules, but there would be no problem with going through compiler transducers and replacing sets of arcs with suitable lists. Finding optimal sets to replace is computationally intractable, but there are some ways to at least do something useful in a reasonable amount of time.)

Yeah that sounds like an excellent idea, I think there will definitely be some common patterns in what kind of lists are usually beneficial.
